File: edition4/js/ch03/3-2.js
var StackNode = function(value, min) {
  this.value = value;
  this.min = min;
}

var Stack = function() {
  this.buffer = [];
}

Stack.prototype.push = function(value) {
  var min = Math.min(value, this.min());
  this.buffer.push(new StackNode(value, min));
}

Stack.prototype.pop = function() {
  return this.buffer.pop();
}

Stack.prototype.min = function() {
  if (this.buffer.length <= 0) {
    return Infinity;
  } else {
    return this.buffer[this.buffer.length - 1].min;
  }
}

Stack.prototype.peek = function() {
  return this.buffer[this.buffer.length - 1];
}

// tests
var assert = require('assert');

describe('Stack', function() {

  describe('push', function() {

    it('should insert correctly', function() {
      var s = new Stack();
      s.push(1);
      s.push(2);
      assert.equal(s.peek().value, 2);
      s.push(3);
      assert.equal(s.peek().value, 3);
    });

  });

  describe('pop', function() {

    it('should pop correctly', function() {
      var s = new Stack();
      s.push(1);
      s.push(2);
      assert.equal(s.peek().value, 2);
      s.push(3);
      assert.equal(s.pop().value, 3);
      assert.equal(s.peek().value, 2);
    });

  });

  describe('min', function() {

    it('should find min correctly', function() {
      var s = new Stack();
      assert.equal(s.min(), Infinity);
      s.push(12);
      assert.equal(s.min(), 12);
      s.push(30);
      assert.equal(s.min(), 12);
      s.push(3);
      assert.equal(s.min(), 3);
    });

  });

  describe('peek', function() {

    it('should peek correctly', function() {
      var s = new Stack();
      s.push(1);
      assert.equal(s.peek().value, 1);
      s.push(2);
      assert.equal(s.peek().value, 2);
    });

  });

});

File: edition4/c/ch03/3-2.cpp
#include <assert.h>

#include <algorithm>
#include <climits>
#include <cmath>
#include <iostream>
#include <stack>
#include <string>

using namespace std;

const int STACK_SIZE = 300;

class MinStack {
  public:
    MinStack();
    void push(int value);
    int pop();
    int min();

  private:
    int buffer[STACK_SIZE];
    int top;
    stack<int> _min;
};

MinStack::MinStack() {
  this->top = 0;
}

void MinStack::push(int value) {
  int top = this->_min.empty() ? INT_MAX : this->_min.top();
  this->_min.push( std::min(top, value) );

  this->buffer[this->top++] = value;
}

int MinStack::pop() {
  this->_min.pop();
  return this->buffer[this->top--];
}

int MinStack::min() {
  return this->_min.empty() ? INT_MAX : this->_min.top();
}

int main() {
  // tests
  MinStack s;

  assert(s.min() == INT_MAX);

  s.push(2);
  s.push(3);
  s.push(1);
  s.push(4);
  s.push(5);

  assert(s.min() == 1);

  return 0;
}
